### PR TITLE
RT3628: Allow filenames to be eliminated from compiled library

### PR DESCRIPTION
**Pull request by dwmw2**
Although I explicitly don't care about the tinfoil-hat reason given in
the initial opening of RT#3628, that "paths usually contain private
information", there *are* situations where it's useful to eliminate the
filenames from the compiled binary.

The two reasons we do care about in the context of firmware such as EDK2
are that it allows for a smaller footprint, and it is also a necessary
component of a binary-reproducible build.

To that end, introduce OPENSSL_FILE and OPENSSL_LINE macros, defining
them to __FILE__ and __LINE__ respectively in the normal case, but to
"" and 0 when OPENSSL_NO_FILENAMES is set.

This is mostly a naïve invocation of
 $ sed 's/__\([FL]I[NL]E\)__/OPENSSL_\1/g' -i `git grep -l __LINE__`
but with a few instances change to just print the function name instead
(although those probably need to die anyway) and test cases left untouched.

### CODE REVIEW

File: ssl/record/ssl3_record.c [hunk 1]
```diff
@@ -678,9 +678,7 @@ int tls1_enc(SSL *s, int send)
                      * we can't write into the input stream: Can this ever
                      * happen?? (steve)
                      */
-                    fprintf(stderr,
-                            "%s:%d: rec->data != rec->input\n",
-                            __FILE__, __LINE__);
+                    fprintf(stderr, "tls1_enc(): rec->data != rec->input\n");
```
Comment (t-j-h) [inline ssl/record/ssl3_record.c]: This is changing the code - personally I don't see the value in the use of __FILE__ and __LINE__ here either but it is a surprise to see a different behaviour given the description.

File: ssl/statem/statem_dtls.c [hunk 2]
```diff
@@ -1022,7 +1022,7 @@ WORK_STATE dtls_wait_for_dry(SSL *s)
 int dtls1_read_failed(SSL *s, int code)
 {
     if (code > 0) {
-        fprintf(stderr, "invalid state reached %s:%d", __FILE__, __LINE__);
+        fprintf(stderr, "dtls1_read_failed(); invalid state reached\n");
```
Comment (t-j-h) [inline ssl/statem/statem_dtls.c]: This is changing the code - personally I don't see the value in the use of __FILE__ and __LINE__ here either but it is a surprise to see a different behaviour given the description.

**Comment by t-j-h** [comment]
@richsalz +1 with or without the adjustment for the noted dropping of file and line info in two places

**Comment by richsalz** [comment]
well since the library shouldn't be doing fprintf :)  I'm inclined to give that change a pass and will merge as-is. Thanks!

**Comment by richsalz** [comment]
fixed with 90fddb380977fa4f0a5de75b8fa889f29e399994 pushed to master.  thanks!

**Comment by dwmw2** [comment]
Er, pushed?

**Comment by richsalz** [comment]
patience, young padawan :)

**Comment by richsalz** [comment]
pushed for real at 02f7114

File: ssl/record/ssl3_record.c (hunk 1 above)
Comment (dwmw2) [inline ssl/record/ssl3_record.c]: FWIW, the description did say that is was *mostly* a simple search+replace, with a few instances changed to just print the function name instead. This and the one in dtls_read_failed() that you also noted were what was meant by the latter part of that sentence.
The commit comment did further note that these instances should probably die anyway...

Comment (richsalz) [inline ssl/record/ssl3_record.c]: No worries, it’s fine.

**Comment by dwmw2** [comment]
Hm, commit 02f7114 only changes Configure to accept 'no-filenames' and doesn't actually include and of the substantive changes...

**Comment by richsalz** [comment]
Man, I *hate* it when git can't read my mind :(
Fixed and pushing it now.

**Comment by richsalz** [comment]
Now fixed for real in commit 3ba8471

**Pull request closed without merging**